File: shared/tests/test_alias_resolver.py
"""Run: python shared/tests/test_alias_resolver.py  (from repo root). No pytest needed."""
import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.dirname(HERE))
from alias_resolver import AliasResolver, FighterRef, normalize, token_sort_ratio, alias_rows_for_fighter  # noqa: E402

FX = json.load(open(os.path.join(HERE, "fixtures.json"), encoding="utf-8"))
failures = 0


def check(cond, msg):
    global failures
    if not cond:
        failures += 1
        print("FAIL:", msg)


for raw, expected in FX["normalize"]:
    got = normalize(raw)
    check(got == expected, f"normalize({raw!r}) = {got!r}, expected {expected!r}")

for a, b, expected in FX["token_sort_ratio"]:
    got = token_sort_ratio(a, b)
    check(abs(got - expected) < 1e-9, f"token_sort_ratio({a!r},{b!r}) = {got}, expected {expected}")

fighters = [FighterRef(id=f["id"], ufcstats_id=f["ufcstats_id"], name=f["name"], nickname=f.get("nickname"),
                       dob=f.get("dob"), record=f.get("record"), weight_classes=set(f.get("weight_classes", [])))
            for f in FX["fighters"]]
r = AliasResolver(fighters)

for case in FX["resolve"]:
    name, source, kw = case["args"]
    res = r.resolve(name, source, **kw)
    exp = case["expect"]
    check(res.status == exp["status"], f"[{case['name']}] status={res.status} expected {exp['status']}")
    if "fighter_id" in exp:
        check(res.fighter_id == exp["fighter_id"], f"[{case['name']}] fighter_id={res.fighter_id} expected {exp['fighter_id']}")
    if "method" in exp:
        check(res.method == exp["method"], f"[{case['name']}] method={res.method} expected {exp['method']}")
    if "n_candidates" in exp:
        n = len(res.review_row["candidate_fighter_ids"]) if res.review_row else 0
        check(n == exp["n_candidates"], f"[{case['name']}] n_candidates={n} expected {exp['n_candidates']}")
    if "reason" in exp:
        got = res.review_row["context"]["reason"] if res.review_row else None
        check(got == exp["reason"], f"[{case['name']}] reason={got} expected {exp['reason']}")

dups = r.find_internal_duplicates()
kinds = {}
for d in dups:
    kinds[d["kind"]] = kinds.get(d["kind"], 0) + 1
exp = FX["internal_duplicates"]
for k, v in exp["expect_kinds"].items():
    check(kinds.get(k, 0) == v, f"internal_duplicates kind {k}: {kinds.get(k, 0)} expected {v} (got {dups})")
near = [d for d in dups if d["kind"] == "near_duplicate"]
check(bool(near) and sorted(near[0]["fighter_ids"]) == sorted(exp["expect_near_pair"]), f"near pair {near}")

rows = alias_rows_for_fighter("f-so", "Sean O'Malley", "Sugar")
check(len(rows) == 2 and rows[0]["normalized"] == "sean omalley" and rows[1]["source"] == "ufcstats_nickname", f"alias rows {rows}")
rows = alias_rows_for_fighter("f-x", "Bruno Silva", None)
check(len(rows) == 1, f"alias rows no nickname {rows}")

print("alias_resolver.py:", "OK" if failures == 0 else f"{failures} FAILURES")
sys.exit(1 if failures else 0)

File: shared/tests/test_alias_resolver.mjs
/* Run: node shared/tests/test_alias_resolver.mjs  (from repo root). */
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { AliasResolver, normalize, tokenSortRatio, aliasRowsForFighter } from '../alias_resolver.mjs';

const HERE = dirname(fileURLToPath(import.meta.url));
const FX = JSON.parse(readFileSync(join(HERE, 'fixtures.json'), 'utf8'));
let failures = 0;
const check = (cond, msg) => { if (!cond) { failures += 1; console.log('FAIL:', msg); } };

for (const [raw, expected] of FX.normalize) {
  const got = normalize(raw);
  check(got === expected, `normalize(${JSON.stringify(raw)}) = ${JSON.stringify(got)}, expected ${JSON.stringify(expected)}`);
}
for (const [a, b, expected] of FX.token_sort_ratio) {
  const got = tokenSortRatio(a, b);
  check(Math.abs(got - expected) < 1e-9, `tokenSortRatio(${a},${b}) = ${got}, expected ${expected}`);
}

const r = new AliasResolver(FX.fighters);
for (const c of FX.resolve) {
  const [name, source, kw] = c.args;
  const res = r.resolve(name, source, kw);
  const exp = c.expect;
  check(res.status === exp.status, `[${c.name}] status=${res.status} expected ${exp.status}`);
  if (exp.fighter_id) check(res.fighter_id === exp.fighter_id, `[${c.name}] fighter_id=${res.fighter_id} expected ${exp.fighter_id}`);
  if (exp.method) check(res.method === exp.method, `[${c.name}] method=${res.method} expected ${exp.method}`);
  if (exp.n_candidates !== undefined) {
    const n = res.review_row ? res.review_row.candidate_fighter_ids.length : 0;
    check(n === exp.n_candidates, `[${c.name}] n_candidates=${n} expected ${exp.n_candidates}`);
  }
  if (exp.dob_conflict !== undefined) check(res.dob_conflict === exp.dob_conflict, `[${c.name}] dob_conflict=${res.dob_conflict} expected ${exp.dob_conflict}`);
  if (exp.reason) {
    const got = res.review_row ? res.review_row.context.reason : null;
    check(got === exp.reason, `[${c.name}] reason=${got} expected ${exp.reason}`);
  }
}

const rows = aliasRowsForFighter('f-so', "Sean O'Malley", 'Sugar');
check(rows.length === 2 && rows[0].normalized === 'sean omalley' && rows[1].source === 'ufcstats_nickname', `alias rows ${JSON.stringify(rows)}`);

console.log('alias_resolver.mjs:', failures === 0 ? 'OK' : `${failures} FAILURES`);
process.exit(failures ? 1 : 0);
